### ISSUE TITLE
Question... Is this possible  to replace Tensorflow with Microsoft CNTK?

### ISSUE THREAD

**Issue by xyhan88** [Zeta36/chess-alpha-zero#62]
Great stuff! I got work in both Windows  and Ubuntu...  Now I get a question to ask, Since it write in Keras which should should support both Google Tensorflow and MS CNTK. Is any simple way to replace Tensorflow with CNTK?

**Comment by brianprichardson** [comment]
Depends on what you mean by simple.  It is like when someone says "if you use your imagination, it tastes like chicken", to which the reply is how much imagination :)

Search for keras in the repo and there are several places where the keras backend is used that would have to be changed I suspect.  While I have never worked with CNTK, it seems possible.

However, I am a bit curious as to why you might want to.

**Comment by xyhan88** [comment]
Took a look at keras menu and played around it a bit... Now I switched keras background to CNTK. It works. It seems it is a bit fast ruining at my Laptop (  I only can run CPU version on laptop)...
Why am I going to try CNTK?  Well I let make comparison between Tensorflow and CNTK about their performance...
BTW,  I saw you have several projects on Repo... So what is the difference between chess-alpha-zero and connect4-alpha-zero? which one is better?

**Comment by brianprichardson** [comment]
Glad it works.  I started with Theano and have been moving to TensorFlow.  Not ready to add CNTK yet.

I am not the owner of this repo, just trying to learn from it.  Connect4 is another game.  In fact, there are quite a few Alpha Zero projects with various games besides Chess and Go.

**Comment by brianprichardson** [comment]
@xyhan88 Please let us know what you had to change for CNTK.
Installing it under Win7 was pretty painless and some of the demos seem to run ok for me, so I thought it would be interesting to see how much faster than TF it is.
Thanks.

**Comment by xyhan88** [comment]
First, you need to install the CNTK. See the following fro MS. I installed CNTK 2.4
https://docs.microsoft.com/en-us/cognitive-toolkit/Setup-CNTK-on-your-machine

The simplest way to switch between Tensorflow and CNTK is to modify keras's settings at .keras/keras.json file. See the following link from Keras:
https://keras.io/backend/

In the keras.json file, just  replace "tensorflow" with "cntk", That is.

Another way is to change the Chess_zero's code and assign keras's backend as cntk. i.e. in model_chess.py file.  But i don't thing it is best way to do it...

In my laptop without using GPU card,  In average, It  takes about 60s plus for every move with tensorflow backend playing in Arena. CNT backend takes less 50s per move...

**Comment by brianprichardson** [comment]
Duh, should have just tried it first.  I had assumed a bunch of changes would need to be made in chess-alpha-zero. 

I commented out the pyperclip references and forced the cntk device to either cpu() or gpu(0) in manager.py per below.  Just to be safe, I re-build a new random best model, and things seems to work.  

I may look at more demanding performance scenarios with CNTK as making a move only takes about 8s on  my slow gpu (a GTX 770), although it is also about 50s on my old Q6600 quad at 2.4GHz.  Tensorflow gpu version is also about 8s.  

The training (optimize.py) time is quite important for me, so I tried that, although I seem to recall speed differences between TF and CNTK can vary quite a bit with batch size, so FWIW:  TF 26s, CNTK 112s in one small test case.

import cntk as C
print(C.device.all_devices())
print(C.device.try_set_default_device(C.device.gpu(0)))
print(C.device.use_default_device())
